doc/hare-dtm-hld.md

```diff
@@ -0,0 +1,570 @@
+Hare controller interface to invoke Hare functionality explicitly.
+
+#### Transient Failure
+A Motr process is said to be transiently failed if the underlying persistent
+storage is still accessible when the process restarts.
+
+#### Participant
+A motr process from the same cluster that contributes to the dtm recovery of
+an another failed motr process.
+
+## Requirements
+
+### [r.failure.transient]
+Hare must be able to detect and report a transient failure.
+
+### [r.ha.epoch]
+Hare must maintain an epoch corresponding to a Motr process and use it to
+record all the relevant events with respect to that process.
+
+### [r.outgoing.msg.timestamp]
+All outgoing HA messages from Hare must have a timestamp that is consistent
+across the cluster.
+
+### [r.failure.order]
+Hare must maintain and broadcast the events in order which they are received
+with respect to a failed node.
+
+### [r.DTM0.messages]
+Hare must implement DTM0 messages as per the protocol in-order to communicate
+with DTM0 services in Motr.
+
+### [r.DTM.participants]
+Hare must know or should be able to identify the participating nodes in DTM
+recovery with respect to a failed node.
+
+### [r.DTM.spiel.interface]
+Hare invokes motr spiel api to send outgoing DTM messages.
+
+### [r.DTM.recover]
+Hare must be able to trigger DTM recovery for a failed node. DTM recovery
+triggers must be sent to the participating nodes only.
+
+### [r.DTM.trigger.once]
+Hare must trigger DTM recovery for a given node exactly once with respect to
+a failure. This requires all the hax process to coordinate.
+
+### [r.DTM.recovery.states]
+Hare must maintain DTM recovery states for all the nodes participating in
+DTM recovery.
+
+### [r.DTM.complete]
+Hare must wait for DTM recovery complete messages from all the participating
+nodes.
+
+### [r.notify.online]
+Hare must broadcast ONLINE ha event for the failed node on successful
+completion of DTM recovery.
+
+### [r.concurrent.failure]
+Hare must handle Hax, Consul and additional Motr process failures during
+an ongoing recovery.
+
+## Design Highlights
+Motr DTM helps to recover an incomplete meta data update transaction that was
+partially executed only on a subset of nodes due a participating process
+failure. Hare monitors and maintains the states for motr processes and is thus
+responsible to trigger DTM recovery in case of transient failures
+i.e. motr process restart, crash, node restart, etc. The recovery is triggered
+when the failed motr process, mainly that hosts ioservice restarts. There are
+multiple sources through which the Hare can receive a process or node failure
+and online events, viz:- Consul and cortx-ha. It is important that Hare records
+all the state changes with respect to a motr process from its start in order as
+they are received, hare uses Consul for the same. Following sequence diagram
+illustrates a typical happy path for DTM recovery triggered by Hare,
+![img](hare-dtm-happy.png)
```

Comment:
>Add a failvec for current post" 
sorry, `post` is a typo, it needs to be epoch.

>The hax that will reply to this entrypoint request may be not an RC. That's why the fact about the previous failure must be kept in Consul KV.
Analysis means that hax needs to check if there is a previous failure registered and if yes, TRANSIENT state must be set.

Correct, we need to maintain failure account to facilitate decision making in Hare with respect to recovery and failure handling. Let me add this to the document.

Reply:
>BTW, if the arrows are numbered, it would be much simpler to reference them while reviewing 😉 Please consider "autonumber" plantuml feature.

Done.